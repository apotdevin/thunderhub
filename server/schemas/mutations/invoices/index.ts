import { parsePayment } from './parsePayment';
import { pay } from './pay';
import { createInvoice } from './createInvoice';
import { payViaRoute } from './payViaRoute';

export const invoices = {
  parsePayment,
  pay,
  createInvoice,
  payViaRoute,
};
