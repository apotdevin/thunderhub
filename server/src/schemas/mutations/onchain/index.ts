import { createAddress } from './getAddress';
import { sendToAddress } from './sendToAddress';

export const onChain = {
    createAddress,
    sendToAddress,
};
