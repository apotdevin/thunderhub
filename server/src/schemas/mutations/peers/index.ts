import { addPeer } from './addPeer';
import { removePeer } from './removePeer';

export const peers = {
    addPeer,
    removePeer,
};
