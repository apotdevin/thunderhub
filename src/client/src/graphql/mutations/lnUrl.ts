import { gql } from '@apollo/client';

export const FETCH_LN_URL = gql`
  mutation FetchLnUrl($url: String!) {
    fetchLnUrl(url: $url) {
      ... on WithdrawRequest {
        callback
        k1
        maxWithdrawable
        defaultDescription
        minWithdrawable
        tag
      }
      ... on PayRequest {
        callback
        maxSendable
        minSendable
        metadata
        commentAllowed
        tag
      }
      ... on ChannelRequest {
        tag
        k1
        callback
        uri
      }
    }
  }
`;

export const AUTH_LN_URL = gql`
  mutation AuthLnUrl($url: String!) {
    lnUrlAuth(url: $url) {
      status
      message
    }
  }
`;

export const PAY_LN_URL = gql`
  mutation PayLnUrl($callback: String!, $amount: Float!, $comment: String) {
    lnUrlPay(callback: $callback, amount: $amount, comment: $comment) {
      tag
      description
      url
      message
      ciphertext
      iv
    }
  }
`;

export const WITHDRAW_LN_URL = gql`
  mutation WithdrawLnUrl(
    $callback: String!
    $amount: Float!
    $k1: String!
    $description: String
  ) {
    lnUrlWithdraw(
      callback: $callback
      amount: $amount
      k1: $k1
      description: $description
    )
  }
`;

export const CHANNEL_LN_URL = gql`
  mutation ChannelLnUrl($callback: String!, $k1: String!, $uri: String!) {
    lnUrlChannel(callback: $callback, k1: $k1, uri: $uri)
  }
`;
