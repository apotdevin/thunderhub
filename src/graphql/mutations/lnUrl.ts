import { gql } from '@apollo/client';

export const FETCH_LN_URL = gql`
  mutation FetchLnUrl($url: String!) {
    fetchLnUrl(url: $url) {
      ... on WithdrawRequest {
        callback
        k1
        maxWithdrawable
        defaultDescription
        minWithdrawable
        tag
      }
      ... on PayRequest {
        callback
        maxSendable
        minSendable
        metadata
        commentAllowed
        tag
      }
    }
  }
`;

export const PAY_LN_URL = gql`
  mutation PayLnUrl($callback: String!, $amount: Int!, $comment: String) {
    lnUrlPay(callback: $callback, amount: $amount, comment: $comment) {
      tag
      description
      url
      message
      ciphertext
      iv
    }
  }
`;

export const WITHDRAW_LN_URL = gql`
  mutation WithdrawLnUrl(
    $callback: String!
    $amount: Int!
    $k1: String!
    $description: String
  ) {
    lnUrlWithdraw(
      callback: $callback
      amount: $amount
      k1: $k1
      description: $description
    )
  }
`;
