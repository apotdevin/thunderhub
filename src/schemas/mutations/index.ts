import { channels } from './channels';
import { invoices } from './invoices';
import { onChain } from './onchain';

export const mutation = { ...channels, ...invoices, ...onChain };
