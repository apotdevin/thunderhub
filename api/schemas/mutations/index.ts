import { channels } from './channels';
import { invoices } from './invoices';
import { onChain } from './onchain';
import { peers } from './peers';
import { chat } from './chat';
import { authTokenMutations } from './authToken';

export const mutation = {
  ...channels,
  ...invoices,
  ...onChain,
  ...peers,
  ...chat,
  ...authTokenMutations,
};
