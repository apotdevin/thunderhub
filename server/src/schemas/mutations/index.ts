import { channels } from './channels';
import { invoices } from './invoices';
import { onChain } from './onchain';
import { peers } from './peers';

export const mutation = {
    ...channels,
    ...invoices,
    ...onChain,
    ...peers,
};
