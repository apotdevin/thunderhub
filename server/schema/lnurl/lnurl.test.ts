import testServer from 'server/tests/testServer';
import fetchMock from 'jest-fetch-mock';
import { GraphQLError } from 'graphql';
import { WITHDRAW_LN_URL } from 'src/graphql/mutations/lnUrl';

jest.mock('ln-service');

describe('LNURL Resolvers', () => {
  beforeEach(() => {
    fetchMock.resetMocks();
  });
  describe('getBitcoinPrice', () => {
    test('success', async () => {
      fetchMock.mockResponseOnce(JSON.stringify({ status: 'SUCCESS' }));
      const { mutate } = testServer();

      const res = await mutate({
        mutation: WITHDRAW_LN_URL,
        variables: {
          callback: 'https://domain.com',
          amount: 1000,
          k1: 'random',
          description: 'ln-withdraw',
        },
      });

      expect(res.errors).toBe(undefined);

      expect(fetchMock).toBeCalledWith(
        'https://domain.com?k1=random&pr=boltEncodedRequest',
        undefined
      );
      expect(res).toMatchSnapshot();
    });
    test('success with callback that has query string', async () => {
      fetchMock.mockResponseOnce(JSON.stringify({ status: 'SUCCESS' }));
      const { mutate } = testServer();

      const res = await mutate({
        mutation: WITHDRAW_LN_URL,
        variables: {
          callback: 'https://domain.com?user=123456',
          amount: 1000,
          k1: 'random',
          description: 'ln-withdraw',
        },
      });

      expect(res.errors).toBe(undefined);

      expect(fetchMock).toBeCalledWith(
        'https://domain.com?user=123456&k1=random&pr=boltEncodedRequest',
        undefined
      );
    });
    test('success but not able to withdraw', async () => {
      fetchMock.mockResponseOnce(JSON.stringify({ status: 'ERROR' }));
      const { mutate } = testServer();

      const res = await mutate({
        mutation: WITHDRAW_LN_URL,
        variables: {
          callback: 'https://domain.com',
          amount: 1000,
          k1: 'random',
          description: 'ln-withdraw',
        },
      });

      expect(res.errors).toStrictEqual([
        new GraphQLError('ProblemWithdrawingFromLnUrlService'),
      ]);
    });
    test('failure', async () => {
      fetchMock.mockRejectOnce(new Error('Error'));
      const { mutate } = testServer();

      const res = await mutate({
        mutation: WITHDRAW_LN_URL,
        variables: {
          callback: 'domain.com',
          amount: 1000,
          k1: 'random',
          description: 'ln-withdraw',
        },
      });

      expect(res.errors).toStrictEqual([
        new GraphQLError('ProblemWithdrawingFromLnUrlService'),
      ]);
      expect(res).toMatchSnapshot();
    });
  });
});
