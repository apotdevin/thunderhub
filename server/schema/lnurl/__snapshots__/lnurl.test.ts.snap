// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LNURL Resolvers getBitcoinPrice failure 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: ProblemWithdrawingFromLnUrlService],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`LNURL Resolvers getBitcoinPrice success 1`] = `
Object {
  "data": Object {
    "lnUrlWithdraw": "requestId",
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
